File: app/assets/bundled/js/dialogs.js
// Dialog suppression shim for the CastCodes embedded browser pane.
//
// wry 0.38 does not expose a JS-dialog handler API on macOS (the WKWebView
// `runJavaScript*Panel` UIDelegate methods are not bridged). Without
// intervention, pages that call `window.alert`, `window.confirm`, or
// `window.prompt` either render nothing (WKWebView default with no
// UIDelegate handler) or in some configurations hang waiting for a
// response that never comes.
//
// This shim is injected at page-creation time via
// `WebViewBuilder::with_initialization_script` so it runs before any
// page script. It replaces the three native functions with safe-default
// returns:
//
//   window.alert(msg)   → no-op + console.warn; returns undefined
//   window.confirm(msg) → no-op + console.warn; returns false (deny)
//   window.prompt(msg)  → no-op + console.warn; returns null (cancel)
//
// The deny/cancel returns mirror what users would pick if they reflexively
// dismissed the dialog. Pages get a deterministic answer and don't hang.
//
// A follow-up can wire host-side IPC to surface a non-modal banner so the
// user sees what was suppressed; today the suppression is silent except
// for the console.warn.

(function () {
    "use strict";

    if (window.__castcodes_dialogs_installed__) {
        return;
    }
    window.__castcodes_dialogs_installed__ = true;

    function safeString(value) {
        if (value === undefined || value === null) return "";
        try {
            return String(value);
        } catch (_e) {
            return "[unstringifiable]";
        }
    }

    window.alert = function (message) {
        try {
            console.warn(
                "[castcodes] window.alert suppressed:",
                safeString(message)
            );
        } catch (_e) { /* noop */ }
    };

    window.confirm = function (message) {
        try {
            console.warn(
                "[castcodes] window.confirm suppressed (returning false):",
                safeString(message)
            );
        } catch (_e) { /* noop */ }
        return false;
    };

    window.prompt = function (message, defaultValue) {
        try {
            console.warn(
                "[castcodes] window.prompt suppressed (returning null):",
                safeString(message),
                "default=" + safeString(defaultValue)
            );
        } catch (_e) { /* noop */ }
        return null;
    };
})();

File: app/src/browser/dialogs.rs
//! Page-init shim that suppresses `window.alert/confirm/prompt`.
//!
//! Background: wry 0.38 doesn't bridge WKWebView's `runJavaScript*Panel`
//! UIDelegate methods on macOS, and there's no public `with_*_dialog_handler`
//! on the builder. Without intervention, pages calling these functions can
//! hang or silently fail in unpredictable ways depending on the host
//! WebKit version. The injected shim replaces them with safe-default
//! returns so pages get a deterministic answer (alert→undefined,
//! confirm→false, prompt→null) and the page event loop never stalls.
//!
//! Limitations:
//!   - No host visibility today. A follow-up can route the suppressed
//!     dialog messages through the existing IPC channel (introduced in
//!     #128 for find-in-page) and surface them as a non-modal banner.
//!   - Sites that genuinely require user confirmation (e.g. "are you
//!     sure?" before deletion) fall back to their no-confirmation path.
//!     Pragmatic deny-by-default mirrors how a user would reflexively
//!     dismiss an unwanted dialog.

/// JS shim loaded via `WebViewBuilder::with_initialization_script` so it
/// runs once per page-creation, before any page-author script.
pub(crate) const INIT_SCRIPT: &str = include_str!("../../assets/bundled/js/dialogs.js");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn init_script_is_loaded() {
        // Sanity check the bundled asset wired in.
        assert!(!INIT_SCRIPT.is_empty());
        assert!(INIT_SCRIPT.contains("__castcodes_dialogs_installed__"));
    }

    #[test]
    fn init_script_overrides_all_three_dialog_functions() {
        for fn_name in ["window.alert", "window.confirm", "window.prompt"] {
            assert!(
                INIT_SCRIPT.contains(fn_name),
                "init script doesn't override {fn_name}"
            );
        }
    }

    #[test]
    fn init_script_returns_safe_defaults() {
        // confirm should return false and prompt should return null —
        // the de-facto "user cancelled" answers. The grep checks the
        // literal return statements survive any future refactor.
        assert!(
            INIT_SCRIPT.contains("return false"),
            "confirm doesn't deny by default"
        );
        assert!(
            INIT_SCRIPT.contains("return null"),
            "prompt doesn't cancel by default"
        );
    }
}
